core/crypto/aes/doc.odin

```diff
@@ -0,0 +1,47 @@
+/*
+package aes implements the AES block cipher and some common modes.
+
+
+An example of how to use 'AES-ECB instance' to decrypt binary files
+
+```odin
+
+package foo
+
+import "core:fmt"
+import "core:crypto/aes"
+import "core:os"
+import "core:bytes"
+
+main :: proc() {
+    data, ok := os.read_entire_file("data.bin")
+    if !ok {fmt.eprintln("Error Reading File")}
+
+    key := transmute([]u8)string("aeskey")
+    ctx: aes.Context_ECB
+
+    aes.init_ecb(&ctx, key)
+
+    if len(data) % aes.BLOCK_SIZE != 0 {
+        fmt.eprintln("Error: Data length is not a multiple of AES block size")
+        return
+    }
+
+    // Allocate space for decrypted data
+    srcdata := make([]byte, len(data))
+
+    // Decrypt data in 16-byte blocks
+    for i in 0 ..< len(data) / aes.BLOCK_SIZE {
```

Comment:
I would rename `srcdata` to `plaintext` and rewrite the loop like this:
```cpp
for len(data) > 0 {
   aes.decrypt_ecb(&ctx, plaintext[:aes.BLOCK_SIZE], data[:aes.BLOCK_SIZE])
   data = data[aes.BLOCK_SIZE:]
   plaintext = plaintext[aes.BLOCK_SIZE:]
}
```

Reply:
For some reason, when I rewrite the for loop this way, it gives me no output. I'm still studying how to write more idiomatic Odin code, and I will rewrite it properly soon.

Reply:
I didn't have an encrypted file to test with so I just wrote it here.

And I think I know why. :-) We're advancing the output slice here, which means it's `0` bytes at the end.
Will need to keep a backup of the original slice, too.

```cpp
_plaintext := plaintext // Preserve original slice view of the decrypted data
for len(data) > 0 {
   aes.decrypt_ecb(&ctx, _plaintext[:aes.BLOCK_SIZE], data[:aes.BLOCK_SIZE])
   data = data[aes.BLOCK_SIZE:]
   _plaintext = _plaintext[aes.BLOCK_SIZE:] // Advance second view of the same destination memory
}
```

Reply:
```odin
package foo
import "core:fmt"
import "core:crypto/aes"
import "core:os"
import "core:bytes"

BLOCK_SIZE :: aes.BLOCK_SIZE

main :: proc() {
	data, ok := os.read_entire_file("data.bin")
	if !ok {
		fmt.eprintln("Error Reading File")
		return
	}

	if len(data) % aes.BLOCK_SIZE != 0 {
		fmt.eprintln("Error: Data length is not a multiple of AES block size")
		return
	}

	key := transmute([]u8)string("aeskey")
	ctx: aes.Context_ECB
	aes.init_ecb(&ctx, key)
	// Allocate space for decrypted data
	plaintext := make([]byte, len(data))

	plain := plaintext // Preserve original slice view of the decrypted data
	for len(data) > 0 {
		aes.decrypt_ecb(&ctx, plain[:BLOCK_SIZE], data[:BLOCK_SIZE])
		data  = data[BLOCK_SIZE:]  // Advance encrypted data
		plain = plain[BLOCK_SIZE:] // Advance second view of the same destination memory
	}

	// Output decrypted data
	fmt.println(plaintext)
	os.write_entire_file("rcdecrypted.bin", plaintext)
}
```

Reply:
Does this work?

Reply:
Yes, worked like a charm